### ISSUE TITLE
Open local file from dropbox?

### ISSUE THREAD

**Issue by olibclarke** [uglymol/uglymol#6]
Hi,

This is probably a very silly question as I have next to no experience in HTML - my apologies in advance.

I'd like to use uglymol to provide a link to a self contained web viewer for a locally stored PDB and map, hosted on a cloud service like dropbox - so I'd like to be able to provide someone with a dropbox link to a uglymol html file located on my dropbox, that automatically loads a pdb and map for viewing from the same dropbox folder.

One use case for this is so I can put a QR code on a poster that allows viewers with mobile devices to quickly inspect the map/model using uglymol.

I figured this ought to be possible by providing a direct download link from dropbox in the "load_pdb" or "load_map" function, but no matter what I try (e.g. see screenshot) it says file not found. I figure I am doing something very stupid but I'm not sure what!

Cheers
Oli

<img width="501" alt="image" src="https://user-images.githubusercontent.com/19766818/65731280-1940c400-e093-11e9-8cf7-8e237f52d04a.png">

**Comment by wojdyr** [comment]
Hi Oliver,
does it work with any other html file?
From what I just read dropbox doesn't allow for hosting web pages:
https://webapps.stackexchange.com/questions/16131/may-i-use-dropbox-as-hosting-for-websites-static-files

**Comment by olibclarke** [comment]
Yes it doesn't anymore, i'll need to figure out a different solution for hosting - but that's not the problem here, I was just loading the uglymol html locally, and providing the pdb as a dropbox link. I couldn't even get it to find a local pdb from my harddrive by entering the path, which makes me think I'm doing something fundamentally wrong...

**Comment by olibclarke** [comment]
In an ideal world I'd like to be able to provide a link to an uglymol session where the user can inspect my map and model, but not download/export them (e.g. by inspecting the html source to find a dropbox link). Is this possible, or am I barking up the wrong tree? 

I really like the viewer, and I'd like to use it for sharing map/model fit when presenting a poster, or when sharing data with colleagues who may not have coot/chimera/etc.

**Comment by wojdyr** [comment]
If the html file is not on the same host, you may get such an error (in web browser's console):
> Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at ... (Reason: CORS header 'Access-Control-Allow-Origin' missing).
Using resources from other hosts is possible only if the server is configured to allow it (PDBe and RCSB are, dropbox isn't).

Loading files from local disk is not possible in most of browsers for security reasons. In Chrome, you'd need to start it with option `--allow-file-access-from-files`. The usual practice in web development is to have a local web server and access files through it. For example, the command `python3 -m http.server` will serve files from the current directory.

Preventing users from downloading the files would be hard. The only thing that comes to my mind is to show only a fragment of the model and map.

**Comment by olibclarke** [comment]
Ah, I see. I don't see why loading a file from a dropbox download link doesn't work though, as in the screenshot? Isn't that similar to loading from the pdb, that is it is not loading from local disk, but from the cloud?

**Comment by olibclarke** [comment]
Basically I'd like something like the Uglymol demo (https://uglymol.github.io/view/), but with the capacity to then share a session as a web link, with the capacity to lock the center of rotation to a particular position (and perhaps lock the zoom to a certain range as well), to give an interactive view of just one part of the structure, e.g. a ligand binding site.

**Comment by wojdyr** [comment]
For security reasons, browsers restrict cross-origin HTTP requests initiated from scripts. Accessing file from PDBe works, because these files are served with CORS headers that explicitly allow using them from scripts on different websites. But it won't work with dropbox.

Maybe you could host it on github (GitHub Pages) or somewhere else?

Locking the center would require some work. If you want to show only part of the structure, maybe creating a pdb file with selected residues would work. The same with the map, it could be limited to selected area.
